File: backend/app/Models/ContratProprietaire.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Storage;

/** Un contrat signe avec un proprietaire, pour l'un de ses appartements. */
class ContratProprietaire extends Model
{
    protected $table = "contrats_proprietaires";

    protected $fillable = [
        "owner_id", "realestate_id", "titre", "date_debut", "date_fin",
        "fichier", "nom_fichier", "type_mime", "taille", "created_by",
    ];

    protected $casts = ["date_debut" => "date", "date_fin" => "date"];

    public function bien()
    {
        return $this->belongsTo(Realstate::class, "realestate_id");
    }

    public function auteur()
    {
        return $this->belongsTo(Manager::class, "created_by");
    }

    public function versTableau(): array
    {
        return [
            "id"            => $this->id,
            "proprietaireId" => $this->owner_id,
            "bienId"        => $this->realestate_id,
            "bien"          => $this->bien?->title,
            "titre"         => $this->titre,
            "dateDebut"     => $this->date_debut?->toDateString(),
            "dateFin"       => $this->date_fin?->toDateString(),
            "url"           => Storage::disk("public")->url($this->fichier),
            "nomFichier"    => $this->nom_fichier,
            "typeMime"      => $this->type_mime,
            "taille"        => $this->taille,
            "creeLe"        => $this->created_at?->toISOString(),
            "par"           => $this->auteur
                ? trim(($this->auteur->first_name ?? "") . " " . ($this->auteur->last_name ?? ""))
                : null,
        ];
    }
}

File: mobile/immobilier-sakane/lib/models/contrat_proprietaire.dart
/// Un contrat signé avec un propriétaire, joint pour l'un de ses appartements.
class ContratProprietaire {
  final int id;
  final int? proprietaireId;
  final int? bienId;
  final String? bien;
  final String? titre;
  final DateTime? dateDebut;
  final DateTime? dateFin;
  final String url;
  final String? nomFichier;
  final String? typeMime;
  final int? taille;
  final DateTime? creeLe;
  final String? par;

  const ContratProprietaire({
    required this.id,
    this.proprietaireId,
    this.bienId,
    this.bien,
    this.titre,
    this.dateDebut,
    this.dateFin,
    required this.url,
    this.nomFichier,
    this.typeMime,
    this.taille,
    this.creeLe,
    this.par,
  });

  bool get estPdf =>
      (typeMime ?? '').contains('pdf') || url.toLowerCase().endsWith('.pdf');

  factory ContratProprietaire.fromJson(Map<String, dynamic> json) {
    return ContratProprietaire(
      id: (json['id'] as num?)?.toInt() ?? 0,
      proprietaireId: (json['proprietaireId'] as num?)?.toInt(),
      bienId: (json['bienId'] as num?)?.toInt(),
      bien: json['bien']?.toString(),
      titre: json['titre']?.toString(),
      dateDebut: DateTime.tryParse((json['dateDebut'] ?? '').toString()),
      dateFin: DateTime.tryParse((json['dateFin'] ?? '').toString()),
      url: (json['url'] ?? '').toString(),
      nomFichier: json['nomFichier']?.toString(),
      typeMime: json['typeMime']?.toString(),
      taille: (json['taille'] as num?)?.toInt(),
      creeLe: DateTime.tryParse((json['creeLe'] ?? '').toString())?.toLocal(),
      par: json['par']?.toString(),
    );
  }
}
